plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchBootableJarMojo.java

```diff
@@ -293,6 +338,131 @@ ModelNode waitStatus(ModelControllerClient client, String status, ModelNode op)
         }
     }
 
+    private class RemoteDeploymentController implements DeploymentController {
+        private final String name;
+
+        public RemoteDeploymentController(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public void deploy(Path dir) throws Exception {
+            try (ModelControllerClient client = createClient()) {
+                ServerHelper.waitForStandalone(client, timeout);
+                undeploy(client, name);
+                waitRemoved(client, name);
+                boolean success = deploy(client, dir);
+                if (success) {
+                    waitDeploymentUp(client, name);
+                }
+                // We only need this on Windows since it may lock the JAR when the delete process is running
+                if (IS_WINDOWS) {
+                    currentServerDir = getHomeDirectory(client);
+                }
+            }
+        }
+
+        private void waitRemoved(ModelControllerClient client, String name) throws Exception {
+            ModelNode address = new ModelNode();
+            address.add("deployment", name);
+            waitStatus(client, "failed", Operations.createOperation("read-resource", address));
+            getLog().debug("Deployment " + name + " removed");
+        }
+
+        private void waitDeploymentUp(ModelControllerClient client, String name) throws Exception {
+            ModelNode address = new ModelNode();
+            address.add("deployment", name);
+            ModelNode op = Operations.createOperation("read-attribute", address);
+            op.get("name").set("status");
+            ModelNode reply = waitStatus(client, "success", op);
+            ModelNode result = reply.get("result");
+            if (result.isDefined()) {
+                String status = result.asString();
+                if ("OK".equals(status)) {
+                    getLog().debug("Deployment " + name + " is up");
+                } else {
+                    getLog().warn("Deployment " + name + " failed, status is " + status);
+                }
+            } else {
+                throw new MojoExecutionException("No status returned for deployment " + name);
+            }
+        }
+
+       private void undeploy(ModelControllerClient client, String name) throws Exception {
```

Comment:
I think that we should share the logic with the localDeploymentController, introduce an AbstractDeploymentController?